File: spec/system/debugging_case_flow_spec.rb
# frozen_string_literal: true

require "rails_helper"

RSpec.describe "Debugging case user flow", type: :system do
  let(:email) { "flow-#{SecureRandom.hex(4)}@example.com" }
  let(:raw_email) { "customer-#{SecureRandom.hex(4)}@secret.example" }
  let(:raw_token) { "sk-flow-#{SecureRandom.hex(8)}" }
  let(:shared_request_id) { "req-flow-#{SecureRandom.hex(6)}" }

  before do
    sign_up_via_browser(email: email)
    expect(page).to have_text("Signed in as #{email}.")
  end

  it "creates a multi-source case, analyzes, exports, and archives from the browser" do
    click_link "New case"
    expect(page).to have_css("h1", text: "New debugging case")

    fill_in "Title", with: "E2E checkout failure"
    fill_in "Customer reference", with: "Contact #{raw_email}"

    fill_log_source_slot(
      1,
      source_type: "Rails log",
      name: "Rails",
      pasted_content: "User login failed for #{raw_email}\nAuthorization: Bearer #{raw_token}"
    )
    fill_log_source_slot(
      2,
      source_type: "Aws cloudwatch",
      pasted_content: "Timeout for request_id=#{shared_request_id}"
    )
    fill_log_source_slot(
      3,
      source_type: "Browser console",
      pasted_content: "Error for request_id=#{shared_request_id}"
    )

    click_button "Create debugging case"

    expect(page).to have_css("h1", text: "E2E checkout failure")
    expect(page).not_to have_text(raw_email)
    expect(page).not_to have_text(raw_token)
    expect(page).not_to have_text(shared_request_id)
    expect(page).to have_text("[EMAIL_1]")
    expect(page).to have_text("[AUTH_1]")
    expect(page).to have_css("h2", text: "Redaction summary")
    within(find("section.card", text: "Sanitized log sources")) do
      expect(page).to have_css("textarea", text: /\[EMAIL_1\]/)
      expect(page).to have_css("textarea", text: /\[AUTH_1\]/)
    end

    click_button "Analyze case"
    expect(page).to have_text("Analysis complete.")
    expect(page).to have_css("h2", text: "Hypothesis report")
    expect(page).to have_css("h2", text: "Correlation signals")
    expect(page).to have_text("[REQUEST_1]")
    expect(page).to have_link("download the report")
    within(find("section.card", text: "Hypothesis report")) do
      expect(page).to have_css("textarea[aria-label='Report Markdown']", text: /## Hypothesis report/)
    end

    debugging_case = DebuggingCase.last
    visit download_report_debugging_case_path(debugging_case)
    expect(page.body).to include("## Hypothesis report")
    expect(page.body).not_to include(raw_email)
    expect(page.body).not_to include(raw_token)

    visit debugging_case_path(debugging_case)
    click_button "Archive case"
    expect(page).to have_text("Case archived.")
    expect(page).to have_css("h1", text: "Debugging cases")

    click_link "Archived"
    expect(page).to have_link("E2E checkout failure")
    expect(page).not_to have_text("No archived debugging cases")

    click_link "Active"
    expect(page).to have_text("No active debugging cases yet")
  end
end

File: e2e/debugging-case-flow.spec.ts
import { test, expect } from "@playwright/test";
import { fillLogSourceSlot, signUp, uniqueEmail } from "./helpers";

test("full debugging case journey in the browser", async ({ page }) => {
  const email = uniqueEmail("pw-flow");
  const rawEmail = uniqueEmail("customer-secret");
  const rawToken = `sk-pw-flow-${Math.random().toString(36).slice(2, 10)}`;
  const sharedRequestId = `req-pw-flow-${Math.random().toString(36).slice(2, 8)}`;

  await signUp(page, email);

  await page.getByRole("link", { name: "New case" }).click();
  await expect(page.getByRole("heading", { name: "New debugging case" })).toBeVisible();

  await page.getByLabel("Title").fill("Playwright checkout failure");
  await page.getByLabel("Customer reference").fill(`Contact ${rawEmail}`);

  await fillLogSourceSlot(page, 1, {
    sourceType: "Rails log",
    name: "Rails",
    pastedContent: `User login failed for ${rawEmail}\nAuthorization: Bearer ${rawToken}`,
  });
  await fillLogSourceSlot(page, 2, {
    sourceType: "Aws cloudwatch",
    pastedContent: `Timeout for request_id=${sharedRequestId}`,
  });
  await fillLogSourceSlot(page, 3, {
    sourceType: "Browser console",
    pastedContent: `Error for request_id=${sharedRequestId}`,
  });

  await page.getByRole("button", { name: "Create debugging case" }).click();

  await expect(page.getByRole("heading", { name: "Playwright checkout failure" })).toBeVisible();
  await expect(page.getByText(rawEmail)).toHaveCount(0);
  await expect(page.getByText(rawToken)).toHaveCount(0);
  await expect(page.getByText(sharedRequestId)).toHaveCount(0);
  await expect(page.getByRole("heading", { name: "Redaction summary" })).toBeVisible();

  const sanitizedSection = page.locator("section.card").filter({ hasText: "Sanitized log sources" });
  const railsLog = sanitizedSection.getByRole("textbox", { name: "Sanitized log for Rails" });
  await expect(railsLog).toContainText("[EMAIL_1]");
  await expect(railsLog).toContainText("[AUTH_1]");

  await page.getByRole("button", { name: "Analyze case" }).click();
  await expect(page.getByText("Analysis complete.")).toBeVisible();
  await expect(page.getByRole("heading", { name: "Hypothesis report" })).toBeVisible();
  await expect(page.getByRole("heading", { name: "Correlation signals" })).toBeVisible();
  const correlationSection = page.locator("section.card").filter({
    has: page.getByRole("heading", { name: "Correlation signals" }),
  });
  await expect(correlationSection).toContainText("[REQUEST_1]");

  const reportMarkdown = page.getByRole("textbox", { name: "Report Markdown" });
  await expect(reportMarkdown).toContainText("## Hypothesis report");

  const downloadPromise = page.waitForEvent("download");
  await page.getByRole("link", { name: "download the report" }).click();
  const download = await downloadPromise;
  const downloadBody = await download.createReadStream();
  const chunks: Buffer[] = [];
  for await (const chunk of downloadBody!) {
    chunks.push(Buffer.from(chunk));
  }
  const markdown = Buffer.concat(chunks).toString("utf8");
  expect(markdown).toContain("## Hypothesis report");
  expect(markdown).not.toContain(rawEmail);
  expect(markdown).not.toContain(rawToken);

  await page.getByRole("button", { name: "Archive case" }).click();
  await expect(page.getByText("Case archived.")).toBeVisible();
  await expect(page.getByRole("heading", { name: "Debugging cases" })).toBeVisible();

  await page.getByRole("link", { name: "Archived" }).click();
  await expect(page.getByRole("link", { name: "Playwright checkout failure" })).toBeVisible();

  await page.getByRole("link", { name: "Active" }).click();
  await expect(page.getByText("No active debugging cases yet")).toBeVisible();
});
